text/0015-status-update.md

```diff
@@ -0,0 +1,397 @@
+
+Adding mutability to this design poses challenges. Fundamentally, the problem
+comes down to Rust's memory safety guarantees. Specifically, Rust forbids
+aliasing mutable references. In the zerocopy case, we would want to be able to
+have mutable references to multiple tables at the same time. These tables share
+the same underlying storage, so we cannot just pass out multiple mutable
+references to that storage to our multiple tables.
+
+The zerocopy crate *is* designed with this in mind: this is the reason for the
+API of `LayoutVerified`, which can read off the front of a slice: under the
+covers, it is splitting the pointer, ensuring that it has exclusive access. This
+pattern may cover *some* of our needs, but not all of them, and even where it
+was feasible it would add significant complexity. With the top-level
+tables, for instance, we could verify when loading the font that the reported
+offsets for the tables were all disjoint, and then we could split the underlying
+data into chunks, each of which would be independently mutable; we could either
+do this explicitly (allocating an array of disjoint slices) or via some other
+mechanism, similar to a `RefCell`, where we would keep track at runtime of which regions
+were currently borrowed.
+
+This would work (although it would add significant complexity) in cases where we
+had foreknowledge of the position and size of an object, but I do not believe
+this could work when dealing with tables that host offsets, or at least not in
+the general case. The problem with offsets is that they are not verified ahead
+of time, and they often point to objects that do not have a known size. This
+means there is really no way to know exactly how much memory a table at a given
+offset is going to access. While it is possible to imagine complicated runtime
+schemes for tracking this, it would definitely require allocating additional
+data structures, and would significantly degrade the ergonomics of the API, and
+significantly increase the learning curve.
+
+I've been thinking about this problem quite a bit, and my opinion it is evidence
+of a fundamental tension in the design goals of this project.
+
+#### deciding on goals
+
+I believe that we currently have two conflicting goals. One of the rationale for
+this project is concern about the difficulty of getting developers up to speed
+on HarfBuzz, and how hard it is to add features there without accidentally
+introducing bugs. There is a hope that Rust's safety guarantees can give us a
+code base that is easier to modify and hack on; however there is *also* a desire
+to end up with code that is more-or-less equivalent to the existing C++.
+
+The basic tension is this: the ergonomics and safety advantages that Rust offers
+work by explicitly preventing certain behaviour, such as the aliasing of mutable
+variables. We cannot translate idiomatic C++ directly into Rust and get safety;
+we get safety by ruling out certain patterns.
```

Comment:
Do you have concrete examples where this is causing problems? - the obvious one would have seemed to be HarfBuzz using mmap and reinterpreting bytes everywhere ... and it turns out Rust does that just fine.

Reply:
mmap is actually a reasonable example, and it's a place where Rust and C++ *do* have different semantics. I touch on this elsewhere, but the main problem is that Rust has strong invariants around the aliasing of pointers. With `mmap`, if we are interpreting an `mmap`'d file as a slice of bytes, any mutation of that file elsewhere would be instant-UB. My (admittedly poor) understanding of C++ leads me to understand that this is not necessarily the case there.

Reply:
I don't *think* HB is coded to assume the file can mutate out from under it (@garretrieger correct?), indeed I rather thought we saw crashing bugs when we did manage to ... I want to say truncate the file it was looking at out from under it.

Reply:
Rust makes promises and assumptions that C++ doesn't, but practically there is very little difference with respect to this issue. mmap is fundamentally unsafe and cannot be made safe without changes at the OS level... essentially guaranteed COW semantics for mapped regions. mmap is too useful, so assuming the risk is worthwhile although that choice should be delegated to the user of the library.